Added imagemin to gulp

accraze opened the pull request:
This PR is in reference to #1066 (Grunt/Gulp tasks)

in order to minify images in the build process,
gulp is now given an imagemin task.

lirantal commented:
LGTM but needs a review from gulp users

denzelwamburu commented:
:+1:

ilanbiala commented on package.json:
```diff
@@ -92,6 +92,7 @@
     "gulp-csslint": "~0.1.5",
     "gulp-cssmin": "~0.1.7",
     "gulp-eslint": "~1.0.0",
+    "gulp-imagemin": "2.4.0",
```
~

ilanbiala commented on package.json:
```diff
@@ -104,6 +105,7 @@
     "gulp-sass": "^2.0.3",
     "gulp-uglify": "^1.2.0",
     "gulp-util": "^3.0.6",
+    "imagemin-pngquant": "4.2.0",
```
~

ilanbiala commented:
Which task runs this?

mleanos commented on gulpfile.js:
```diff
@@ -159,6 +161,17 @@ gulp.task('less', function () {
 
+// Imagemin task
+gulp.task('imagemin', function () {
+  return gulp.src(defaultAssets.client.img)
```
Where are these assets coming from? We don't currently have a definition for *client.img*.

mleanos commented on config/assets/default.js:
```diff
@@ -35,6 +35,9 @@ module.exports = {
       'modules/*/client/**/*.js'
     ],
+    img: [
+      'modules/users/client/img/**/*.png'
```
We only want to support minifying *.png*?

mleanos commented:
+1 on adding this to the `build` task.

accraze commented:
thanks for the feedback guys! I added `imagemin` to the `build` task and added support for all img file types.... let me know if anything else needs changing :grin:

ilanbiala commented on config/assets/default.js:
```diff
@@ -35,6 +35,9 @@ module.exports = {
       'modules/*/client/**/*.js'
     ],
+    img: [
+      'modules/**/**/img/**/*'
```
You can still specify specific images, I'd rather not just let this bluntly run on any file in there because there may be user error.

accraze replied:
ok... should we just only handle .png, .jpg, .gif and .ico file types?

ilanbiala replied:
Handle whichever ones the png library supports? Anything else and we may have an error, so we can just avoid that by using the multiple term matching syntax, which I think is `(abc|xyz|def)`, but verify that.

rhutchison commented:
Why does this need to be a part of the build process?  Seems a bit opinionated?  What if I don't care to compress my images?

rhutchison commented on gulpfile.js:
```diff
@@ -11,9 +11,11 @@ var _ = require('lodash'),
     rename: {
-      'gulp-angular-templatecache': 'templateCache'
+      'gulp-angular-templatecache': 'templateCache',
+      'gulp-imagemin': 'imagemin'
```
not required

ilanbiala commented:
@accraze any progress on this?

ilanbiala commented:
@accraze can you finalize this PR?

accraze commented:
sure no prob... did we ever reach a consensus on whether or not to include `imagemin` in the build task? I hear what @rhutchison is saying about it being opinionated... i could go either way... some projects won't really need imagemin though...

Aside from the coveralls issues, imagemin is ready to go and supports jpg, png, gif and svg. Will send a fix later today

mleanos commented:
Let's not add it to the build task. Someone could easily add it to the build sequence if they wish to.

accraze commented:
cool... removed imagemin from the build task, fixed the coveralls issue and squashed all the commits... lemme know if you guys need anything else

mleanos commented:
Looks like there's quite a bit of diffs that have already been merged into master. You'll need to rebase and then squash.

ilanbiala commented:
Are all the new changes because you merged to resolve conflicts?

accraze commented:
yeah sorry I should have mentioned.... I had to merge master in order to get the coveralls tests to pass... it was failing due to missing test coverage.

ilanbiala commented:
@accraze can you try doing what @mleanos mentioned so that we can have only the necessary changes in this PR. Otherwise merging this in makes it look like a lot was changed.

accraze commented:
yeah no prob... will fix later this evening

mleanos commented:
@accraze I'll be on [Gitter meanjs](https://gitter.im/meanjs/mean) most of the day & evening, if you need tips on doing the rebase.

accraze commented:
took the merge commits out with a rebase last night... should be good to go

mleanos commented:
LGTM. @ilanbiala @rhutchison

ilanbiala commented on config/assets/default.js:
```diff
@@ -35,6 +35,12 @@ module.exports = {
       'modules/*/client/**/*.js'
     ],
+    img: [
+      'modules/**/**/img/**/*.jpg',
```
2nd double asterisk is unnecessary, just single asterisk.

ilanbiala commented on gulpfile.js:
```diff
@@ -12,9 +12,10 @@ var _ = require('lodash'),
   plugins = gulpLoadPlugins({
     rename: {
-      'gulp-angular-templatecache': 'templateCache'
+      'gulp-angular-templatecache': 'templateCache',
```
remove comma

ilanbiala commented:
Nice, just 2 fixes then I'll merge.

accraze commented:
@ilanbiala... fixed!
